pkg/controller/unreachable/unreachable_controller_test.go

```diff
@@ -0,0 +1,239 @@
+/*
+Copyright (C) 2019 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package unreachable
+
+import (
+	"context"
+	"errors"
+	"testing"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/openshift/hive/pkg/apis"
+	controllerutils "github.com/openshift/hive/pkg/controller/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes/scheme"
+
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	hivev1 "github.com/openshift/hive/pkg/apis/hive/v1alpha1"
+)
+
+const (
+	testName              = "foo-unreachable"
+	testClusterName       = "bar"
+	testClusterID         = "testFooClusterUUID"
+	installJobName        = "foo-unreachable-install"
+	uninstallJobName      = "foo-unreachable-uninstall"
+	testNamespace         = "default"
+	metadataName          = "foo-unreachable-metadata"
+	adminPasswordSecret   = "admin-password"
+	sshKeySecret          = "ssh-key"
+	pullSecretSecret      = "pull-secret"
+	testUUID              = "fakeUUID"
+	testAMI               = "ami-totallyfake"
+	adminKubeconfigSecret = "foo-unreachable-admin-kubeconfig"
+	adminKubeconfig       = `clusters:
+- cluster:
+    certificate-authority-data: JUNK
+    server: https://bar-api.clusters.example.com:6443
+  name: bar
+`
+	testRemoteClusterCurrentVersion = "4.0.0"
+	remoteClusterVersionObjectName  = "version"
+
+	remoteClusterRouteObjectName      = "console"
+	remoteClusterRouteObjectNamespace = "openshift-console"
+)
+
+func init() {
+	log.SetLevel(log.DebugLevel)
+}
+
+func TestUnreachableClusterDeployment(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+	tests := []struct {
+		name        string
+		existing    []runtime.Object
+		expectError bool
+		validate    func(*testing.T, *hivev1.ClusterDeployment)
```

Comment:
We should have some testing around the last probe and transition times, as we do not want to introduce a hotloop here. i.e. A reconcile when we're within the last probe time limits.

Reply:
Good point. Will fix that.

Reply:
Added the tests.